File: examples-standalone/kendoangular-aspnetcore-integration/Controllers/ProductController.cs

using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using Kendo.Mvc.Extensions;
using Kendo.Mvc.UI;
// To add Telerik NuGet packages, check the Telerik UI for ASP.NET Core documentation: https://docs.telerik.com/reporting/getting-started/installation/adding-private-nuget-feed


namespace Kendo_ASP.NET_Core_Angular.Controllers
{
    [ApiController]
    [Route("products")]
    public class ProductController : ControllerBase
    {
        private readonly ILogger<ProductController> _logger;

        public ProductController(ILogger<ProductController> logger)
        {
            _logger = logger;
        }

        private static List<Product> Products = new List<Product>
        {
            new Product
            {
                ProductID = 1,
                ProductName = "Chai",
                UnitPrice = 18,

            },
            new Product
            {
                ProductID = 2,
                ProductName = "Chang",
                UnitPrice = 19,

            },
            new Product
            {
                ProductID = 3,
                ProductName = "Aniseed Syrup",
                UnitPrice = 10,

            },
            new Product
            {
                ProductID = 4,
                ProductName = "Chef Anton's Cajun Seasoning",
                UnitPrice = 22,

            },
            new Product
            {
                ProductID = 5,
                ProductName = "Chef Anton's Gumbo Mix",
                UnitPrice = 21.35M,

            },
            new Product
            {
                ProductID = 6,
                ProductName = "Grandma's Boysenberry Spread",
                UnitPrice = 25,

            },
            new Product
            {
                ProductID = 7,
                ProductName = "Uncle Bob's Organic Dried Pears",
                UnitPrice = 30,

            },
            new Product
            {
                ProductID = 8,
                ProductName = "Northwoods Cranberry Sauce",
                UnitPrice = 40,

            },
            new Product
            {
                ProductID = 9,
                ProductName = "Mishi Kobe Niku",
                UnitPrice = 97,

            },
            new Product
            {
                ProductID = 10,
                ProductName = "Ikura",
                UnitPrice = 31,

            }
        };

        [HttpPost]
        [Route("get-products")]
        public ActionResult GetProducts([DataSourceRequest] DataSourceRequest request)
        {

            var requestJson = JsonConvert.SerializeObject(request);
            _logger.LogInformation("Data requested: {RequestJson}", requestJson);

            var result = Products.AsQueryable().ToDataSourceResult(request);

            var resultJson = JsonConvert.SerializeObject(result);
            _logger.LogInformation("Data result: {ResultJson}", resultJson);

            return Ok(result);
        }

        [HttpPost]
        [Route("create-product")]
        public IActionResult CreateProduct([FromBody] Product product)
        {
            Products.Add(product);

            _logger.LogInformation("Product created: {Product}", JsonConvert.SerializeObject(product));
            return Ok(product);
        }


        [HttpPut]
        [Route("update-product/{originalProductId}")]
        public IActionResult UpdateProduct(int originalProductId, [FromBody] Product product)
        {
            var existingProduct = Products.FirstOrDefault(p => p.ProductID == originalProductId);
            if (existingProduct == null)
            {
                return NotFound();
            }

            existingProduct.ProductName = product.ProductName;
            existingProduct.UnitPrice = product.UnitPrice;
            existingProduct.ProductID = product.ProductID;

            _logger.LogInformation("Product updated: {Product}", JsonConvert.SerializeObject(existingProduct));
            return Ok(existingProduct);
        }

        [HttpDelete]
        [Route("delete-product/{id:int}")]
        public IActionResult DeleteProduct(int id)
        {
            Console.WriteLine("Product deleted: {0}", id);
            var product = Products.FirstOrDefault(p => p.ProductID == id);
            if (product == null)
            {
                return NotFound();
            }

            Products.Remove(product);
            _logger.LogInformation("Product deleted: {0}", id);

            return Ok();
        }

    }

}

File: examples-standalone/kendoangular-java-integration/backend/src/main/java/telerik/demo/controller/ProductController.java
package telerik.demo.controller;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.bind.annotation.*;

import telerik.demo.datasource.DataSourceRequest;
import telerik.demo.datasource.DataSourceResult;
import telerik.demo.datasource.QueryableExtensions;
import telerik.demo.models.Product;
import telerik.demo.models.ProductDataResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/products")
public class ProductController {

    private static final Logger logger = LoggerFactory.getLogger(ProductController.class);

    private static List<Product> products = new ArrayList<>();

    static {
        products.add(new Product(1, "Chai", 18));
        products.add(new Product(2, "Chang", 19));
        products.add(new Product(3, "Aniseed Syrup", 10));
        products.add(new Product(4, "Chef Anton's Cajun Seasoning", 22));
        products.add(new Product(5, "Chef Anton's Gumbo Mix", 21.35));
        products.add(new Product(6, "Grandma's Boysenberry Spread", 25));
        products.add(new Product(7, "Uncle Bob's Organic Dried Pears", 30));
        products.add(new Product(8, "Northwoods Cranberry Sauce", 40));
        products.add(new Product(9, "Mishi Kobe Niku", 97));
        products.add(new Product(10, "Ikura", 31));
        products.add(new Product(11, "Queso Cabrales", 21));
        products.add(new Product(12, "Queso Manchego La Pastora", 38));
    }

    @GetMapping("/get-products")
    public ProductDataResult getProducts() {
        logger.info("Products requested");
        return new ProductDataResult(products, products.size());
    }

    /**
     * Example using the new DataSourceRequest helper - similar to ASP.NET MVC
     * This endpoint accepts POST with DataSourceRequest in the body
     */
    @PostMapping("/get-products")
    public DataSourceResult getProductsWithDataSource(@RequestBody DataSourceRequest request) {
        logger.info("Products requested with DataSourceRequest: {}", request);
        
        // Use the helper method similar to ASP.NET: products.ToDataSourceResult(request)
        DataSourceResult result = QueryableExtensions.toDataSourceResult(products, request);
        
        logger.info("Returning {} items out of {} total", 
            result.getData() instanceof List ? ((List<?>) result.getData()).size() : "grouped",
            result.getTotal());
        
        return result;
    }

    @PostMapping("/create-product")
    public Product createProduct(@RequestBody Product product) {
        products.add(product);
        logger.info("Product created: {}", product);
        return product;
    }

    @PutMapping("/update-product/{originalProductId}")
    public Product updateProduct(@PathVariable int originalProductId, @RequestBody Product product) {
        Optional<Product> existingProduct = products.stream()
                .filter(p -> p.getProductID() == originalProductId)
                .findFirst();

        if (!existingProduct.isPresent()) {
            logger.warn("Product with ID {} not found", originalProductId);
            return null; // Ideally, throw an exception or return a proper response.
        }

        Product prod = existingProduct.get();
        prod.setProductName(product.getProductName());
        prod.setUnitPrice(product.getUnitPrice());
        prod.setProductID(product.getProductID());

        logger.info("Product updated: {}", prod);
        return prod;
    }

    @DeleteMapping("/delete-product/{id}")
    public void deleteProduct(@PathVariable int id) {
        products = products.stream()
                .filter(p -> p.getProductID() != id)
                .collect(Collectors.toList());

        logger.info("Product deleted: {}", id);
    }
}